middleware/grpc_stats.go

```diff
@@ -0,0 +1,67 @@
+package middleware
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc/stats"
+)
+
+func NewStatsHandler(receivedPayloadSize, sentPayloadSize *prometheus.HistogramVec) stats.Handler {
+	return &grpcStatsHandler{
+		receivedPayloadSize: receivedPayloadSize,
+		sentPayloadSize:     sentPayloadSize,
+	}
+}
+
+type grpcStatsHandler struct {
+	receivedPayloadSize *prometheus.HistogramVec
+	sentPayloadSize     *prometheus.HistogramVec
+}
+
+// Custom type to hide it from other packages.
+type contextKey int
+
+const (
+	contextKeyMethodName contextKey = 1
+)
+
+func (g *grpcStatsHandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) context.Context {
+	return context.WithValue(ctx, contextKeyMethodName, info.FullMethodName)
+}
+
+func (g *grpcStatsHandler) HandleRPC(ctx context.Context, rpcStats stats.RPCStats) {
+	// We use full method name from context, because not all RPCStats structs have it.
+	fullMethodName, ok := ctx.Value(contextKeyMethodName).(string)
+	if !ok {
+		return
+	}
+
+	switch s := rpcStats.(type) {
+	case *stats.Begin:
+		// Ignored.
+	case *stats.End:
+		// Ignored.
+	case *stats.InHeader:
+		// WireLength seems to be incorrect for large headers -- it uses
+		// length of last frame (16K) even for headers in megabytes.
```

Comment:
Please link the gRPC issue number.

Reply:
I've removed the comment instead. I couldn't find gRPC issue. It's something I've observed, but I'm not even sure if it's an issue or not (I would expect WireLength to be higher than message size, but it was just 16K).